tests/func/experiments/test_checkpoints.py

```diff
@@ -188,3 +192,73 @@ def test_resume_non_head_checkpoint(
     )
     new_head = first(results)
     assert orig_branch != dvc.experiments.get_branch_by_rev(new_head)
+
+
+@pytest.mark.parametrize("use_url", [True, False])
+def test_auto_push_during_iterations(
+    tmp_dir, scm, dvc, checkpoint_stage, git_upstream, local_remote, use_url
+):
+    # set up remote repo
+    remote = git_upstream.url if use_url else git_upstream.remote
+    git_upstream.scm.fetch_refspecs(str(tmp_dir), ["master:master"])
+
+    # without auto push
+    results = dvc.experiments.run(checkpoint_stage.addressing)
+    exp = first(results)
+    ref_info = first(exp_refs_by_rev(scm, exp))
+    assert git_upstream.scm.get_ref(str(ref_info)) is None
+
+    # add auto push
+    os.environ["DVC_EXP_AUTO_PUSH"] = remote
+    results = dvc.experiments.run(checkpoint_stage.addressing)
+    assert (tmp_dir / "foo").read_text() == "4"
+    exp = first(results)
+    ref_info = first(exp_refs_by_rev(scm, exp))
+    assert git_upstream.scm.get_ref(str(ref_info)) == exp
+
+    # check the data
+    with git_upstream.dvc.config.edit() as conf:
+        conf["remote"]["local"] = local_remote.config
+        conf["core"]["remote"] = "local"
+
+    git_upstream.dvc.experiments.apply(ref_info.name)
+    git_upstream.dvc.experiments.apply(exp)
+    git_upstream.dvc.pull()
+    assert (git_upstream / "foo").read_text() == "4"
+
+    # resume the remote checkpoint
+    os.environ.pop("DVC_EXP_AUTO_PUSH")
+    with git_upstream.chdir():
+        git_upstream.dvc.experiments.run(checkpoint_stage.addressing)
+    assert (git_upstream / "foo").read_text() == "6"
+
+
+def test_auto_push_error_url(dvc, scm, checkpoint_stage, local_remote):
+    os.environ["DVC_EXP_AUTO_PUSH"] = "none"
```

Comment:
How to test the error message from an exception? None of the  `caplog.text`, `capsys.err` and `capsys.out` show this out besides the exception was inside the executor hard to be tested by `pytest`

Reply:
As you noted, you can't test for exceptions and logging messages here like usual, since experiment execution happens in a separate process.

But any exception that is raised in the experiment will be returned by `future.exception()` and collected in the main process:
https://github.com/iterative/dvc/blob/1e43c303a7bb7c2c34c721554c859ae3eb54c938/dvc/repo/experiments/__init__.py#L710

If you just want to test that a particular exception type generates a specific message, you can just patch the `future.exception()` call so that it returns the exception you want:
https://github.com/iterative/dvc/blob/master/dvc/repo/experiments/__init__.py#L710

You could also patch `future.exception` and check all of the times it is actually called, and look for whatever particular exception you are testing for